Source/SwiftLintBuiltInRules/Rules/Idiomatic/DiscouragedOptionalCollectionRule.swift

```diff
@@ -12,153 +12,66 @@ struct DiscouragedOptionalCollectionRule: ASTRule, OptInRule, ConfigurationProvi
-    ///
-    ///         [  S  t  r  i  n  g  :     [  I  n  t  ]  ?  ]
-    ///         0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15
-    ///                                    ^              ^
-    /// = [9, 14]
-    /// = [9, 15), mathematical interval, w/ lower and upper bounds.
-    ///
-    /// Example: [String: [Int]?]?
-    ///
-    ///         [  S  t  r  i  n  g  :     [  I  n  t  ]  ?  ]  ?
-    ///         0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16
-    ///         ^                          ^              ^     ^
-    /// = [0, 16], [9, 14]
-    /// = [0, 17), [9, 15), mathematical interval, w/ lower and upper bounds.
-    ///
-    /// Example: var x = Set<Int>?
-    ///
-    ///         v  a  r     x     =     S  e  t  <  I  n  t  >  ?
-    ///         0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15 16
-    ///                                 ^                       ^
-    /// = [8, 16]
-    /// = [8, 17), mathematical interval, w/ lower and upper bounds.
-    ///
-    /// - returns: An array of ranges.
-    func optionalCollectionRanges() -> [Range<String.Index>] {
-        let squareBrackets = balancedRanges(from: "[", to: "]").compactMap { range -> Range<String.Index>? in
-            guard
-                range.upperBound < endIndex,
-                let finalIndex = index(range.upperBound, offsetBy: 1, limitedBy: endIndex),
-                self[range.upperBound] == "?" else { return nil }
+private extension DiscouragedOptionalCollectionRule {
+    final class Visitor: ViolationsSyntaxVisitor {
+        override func visitPost(_ node: OptionalTypeSyntax) {
+            if node.wrappedType.isCollectionType {
+                violations.append(node.positionAfterSkippingLeadingTrivia)
+            }
+        }
 
-            return range.lowerBound..<finalIndex
+        override func visitPost(_ node: OptionalChainingExprSyntax) {
+            if node.expression.isCollectionExpression {
+                violations.append(node.positionAfterSkippingLeadingTrivia)
+            }
         }
 
-        let angleBrackets = balancedRanges(from: "<", to: ">").compactMap { range -> Range<String.Index>? in
+        override func visitPost(_ node: FunctionCallExprSyntax) {
             guard
-                range.upperBound < endIndex,
-                let initialIndex = index(range.lowerBound, offsetBy: -3, limitedBy: startIndex),
-                let finalIndex = index(range.upperBound, offsetBy: 1, limitedBy: endIndex),
-                self[initialIndex..<range.lowerBound] == "Set",
-                self[range.upperBound] == "?" else { return nil }
+                let calledExpression = node.calledExpression.as(MemberAccessExprSyntax.self),
+                let base = calledExpression.base?.as(IdentifierExprSyntax.self),
+                base.identifier.text == "Optional",
+                calledExpression.name.text == "some",
+                let expression = node.argumentList.first?.expression
```

Comment:
Honestly, I don't expect people to ever right an optional collection in this way, but it's already there so we should keep it.

Later we can check if removing this case causes any regressions in OSS projects.